### DISCUSSION TITLE
Probabilistic Programming Category?

### DISCUSSION THREAD

**Discussion by AkihisaYamada** [TermCOMP/starexec-master#93] [General]
Let's also discuss about probabilistic programs.
I'm informed of https://github.com/Sheshansh/prob_termination.

An [example there](https://github.com/Sheshansh/prob_termination/blob/master/implementation/examples/probabilistic%20assignments%20and%20while%20loops/aaron2.prob)
```
var tx,x,y;
if tx>=0 then
	while x>=y and tx>=0 do
	if prob(0.3) then
		if * then
			x:=x-1-tx+[-1,1]
		else
			y:=y+1+tx+[-3,2]
		fi
	else skip fi
	od
else
	skip
fi
```
There are probabilistic and nondeterminisitic branching, vars are reals, `[l,r]` is uniform sampling.
I guess a C-like format (or TRS-like one) with these features are good starting point.

**Comment by ffrohn** [comment]
I think we should use an extension of the [format for LTSs](https://github.com/orgs/TermCOMP/discussions/90) with probabilities as in the [format for PTRSs](https://github.com/orgs/TermCOMP/discussions/82) (which would suffice to represent the example above, as far as I can see).

We had [far too many different formats](https://github.com/orgs/TermCOMP/discussions/91) and we're currently trying to unify them, so using a C-like (or entirely different) format would be counterproductive.

I think other potential participants (like KoAT) are more interested in integer-valued variables. Is there more than one potential participant with support for reals, or more than one potential participant with support for integers?

**Reply by AkihisaYamada** [reply]
We need at least uniform sampling (`[-1,1]` in the example). Maybe allow `(uniform (- 1) 1)` in expression?

**Comment by eleanoremeyer** [comment]
Another option would be to use a format similar to that of [probabilistic integer programs](https://aprove-developers.github.io/ExpectedUpperBounds/#input-format) used by our tool [KoAT](https://koat.verify.rwth-aachen.de/).
Probabilistic integer programs generalise [integer transition systems](https://koat.verify.rwth-aachen.de/getting_started/input_format/) which are already [part of the TermComp & TPDB](https://github.com/TermCOMP/TPDB/tree/master/Complexity_ITS) to also allow for the representation of probabilistic branching and sampling from (possibly parametrised & infinite) probability distributions (over the integers).
This is in extension to the concepts of nondeterministic branching and nondeterministic “sampling” (temporary variables).
The format is also similar to the concept of probabilistic control flow graphs as used [in this publication](https://doi.org/10.1145/3585391) (among others).

Another alternative that comes to mind, is a python-like imperative programming language as used by [eco-imp/ev-imp](https://www-sop.inria.fr/members/Martin.Avanzini/software/eco-imp/) and absynth (see [this publication](https://doi.org/10.1145/3192366.3192394)).
However, I fear that the usage of such a higher-level language would increase the entry barrier for tools.
Moreover, unlike as for probabilistic integer programs, this would not be a generalisation of a format that is already included in the termination competition.

**Comment by ffrohn** [comment]
> Probabilistic integer programs generalise [integer transition systems](https://koat.verify.rwth-aachen.de/getting_started/input_format/) which are already [part of the TermComp & TPDB](https://github.com/TermCOMP/TPDB/tree/master/Complexity_ITS)

But they are about to be [replaced](https://github.com/orgs/TermCOMP/discussions/90).

> Another alternative that comes to mind, is a python-like imperative programming language

See my previous comment:

> We had [far too many different formats](https://github.com/orgs/TermCOMP/discussions/91) and we're currently trying to unify them, so using a C-like (or entirely different) format would be counterproductive.

Anyway, before we fix the syntax, we should clarify some more important questions:

* Are there several tools that support the same kind of programs?
  * KoAT supports integers only (?)
  * [The tool mentioned by Akihisa](https://github.com/Sheshansh/prob_termination) supports reals only (?)
* Would they be willing to participate?
* Do we have (sufficiently many) examples?

**Comment by ffrohn** [comment]
It seems like there's not much interest in such a category at the moment -- closed for now.
